### PR TITLE
GitHub Codespaces/code-server support

### PR DESCRIPTION
**Pull request by TamiTakamiya**
For GitHub Codespaces/code-server support for enabling OAuth2 authorization code flow used on Ansible Lightspeed server.  The PoC code worked both on GitHub Codespaces and code-server, but not on Red Hat Openshift Dev Spaces with a 404 error, which was already reported.

### CODE REVIEW

File: src/features/lightspeed/lightSpeedOAuthProvider.ts [hunk 1]
```diff
@@ -86,6 +89,11 @@ export class LightSpeedAuthenticationProvider
     );
   }
 
+  async setExternalRedirectUri() {
+    const callbackUri = await env.asExternalUri(Uri.parse(this.redirectUri));
```
Comment (TamiTakamiya) [inline src/features/lightspeed/lightSpeedOAuthProvider.ts:93]: When VS Code is running as a standalone application, env.asExternalUri() will return the input URI without modifications ([reference](https://code.visualstudio.com/api/references/vscode-api#env.asExternalUri)).

File: src/features/lightspeed/lightSpeedOAuthProvider.ts [hunk 2]
```diff
@@ -349,7 +353,7 @@ export class LightSpeedAuthenticationProvider
       client_id: LIGHTSPEED_CLIENT_ID,
       code: code,
       code_verifier: CODE_VERIFIER,
-      redirect_uri: this.redirectUri,
+      redirect_uri: this._externalRedirectUri,
```
Comment (TamiTakamiya) [inline src/features/lightspeed/lightSpeedOAuthProvider.ts:356]: Here the redirect URI is used again. However, it is not used for redirection and need to be the same value used for redirection to send the authorization code. So `this.setExternalRedirectUri()` is NOT called there and the stored `this._externalRedirectUri` value is set to the `redirect_uri` property.

**Review by ganeshrn — COMMENTED** [review]
@TamiTakamiya Can you please raise this PR from the branch within the repository in order to run the lightspeed e2e test in the CI

**Comment by l0rd** [comment]
FYI, this can be tested on Red Hat Developer Sandbox using a nightly build of VS Code ([that includes this fix](https://github.com/che-incubator/che-code/pull/319)). Here is a [link](https://workspaces.openshift.com/#https://github.com/devspaces-samples/ansible-devspaces-demo/tree/devspaces-3.10-rhel-8?che-editor=https://eclipse-che.github.io/che-plugin-registry/main/v3/plugins/che-incubator/che-code/insiders/devfile.yaml) to start a workspace that uses nightly VS Code.

image

**Review by ganeshrn — APPROVED** [review]

**Comment by ganeshrn** [comment]
@TamiTakamiya Thank you!

**Pull request merged**